### PR TITLE
Implementation of the Float Domain

### PR DESCRIPTION
**Pull request by FelixKrayer**
I implemented most of the FloatDomain and added Unit Tests. 
It currently uses the `round` module from https://github.com/tyabu12/ocaml-round but we will replace this once we have our own implementation. 
Please feel free to suggest further unit tests.
What's left to do is the `invariant` and the `tag` function. For the latter, I was just unable to understand what it is supposed to do.

The reason that so many Actions fail is that either it is unable to build the `round` module or for the Indentation, it compares the indentation with the previous commit for some reason.

### CODE REVIEW

File: src/cdomains/floatDomain.ml [hunk 1]
```diff
@@ -37,130 +38,210 @@ module type FloatDomainBase = sig
...
-  let le _ _ = failwith "todo - le"
-
-  let ge _ _ = failwith "todo - ge"
-
-  let eq = eval_int_binop eq
-
-  let ne = eval_int_binop ne
+  let add op1 op2 = 
+    match op1, op2 with
+    | Some (l1, h1), Some (l2, h2) -> Some (add Down l1 l2, add Up h1 h2)
```
Comment (Dudeldu) [inline src/cdomains/floatDomain.ml]: norm required

File: src/cdomains/floatDomain.ml [hunk 2]
```diff
@@ -37,130 +38,210 @@ module type FloatDomainBase = sig
...
+    | _ -> None
+
+  let sub op1 op2 = 
+    match op1, op2 with
```
Comment (Dudeldu) [inline src/cdomains/floatDomain.ml]: maybe introduce a helper method again, that unwraps the single operand, as this pattern is repeated for every single operation. This could also handle the norm application, that is required for all cases

File: src/cdomains/floatDomain.ml [hunk 3]
```diff
@@ -37,130 +38,210 @@ module type FloatDomainBase = sig
...
+  let lt op1 op2 =
+    let (a, b) =
+      match op1, op2 with
+      | Some (l1, h1), Some (l2, h2) -> 
+        if h1 < l2 then (1, 1)
+        else if l1 >= h2 then (0, 0) 
+        else (0, 1)
+      | _ -> (0, 1)
+    in
+    IntDomain.IntDomTuple.of_interval IBool (Big_int_Z.big_int_of_int a, Big_int_Z.big_int_of_int b)
```
Comment (Dudeldu) [inline src/cdomains/floatDomain.ml]: this wrapping into IntInterval could also be moved into a helper function. For exact results, we could also use `IntDomain.IntDomTuple.of_int` instead of the interval constructor.

File: unittest/cdomains/floatDomainTest.ml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+module FloatInterval =
+struct
+  module FI = FloatDomain.FloatInterval
+  module IT = IntDomain.IntDomTuple
+
+  let fi_zero = FI.of_const 0.
+  let fi_one = FI.of_const 1.
+  let fi_neg_one = FI.of_const (-.1.)
+  let itb_true = IT.of_interval IBool (Big_int_Z.big_int_of_int 1, Big_int_Z.big_int_of_int 1)
```
Comment (Dudeldu) [inline unittest/cdomains/floatDomainTest.ml]: `IT.of_int`

File: unittest/mainTest.ml [hunk 5]
```diff
@@ -10,4 +11,4 @@ let all_tests = ("" >:::
     "domaintest" >::: QCheck_ounit.to_ounit2_test_list Maindomaintest.all_testsuite
   ])
 
-let () = run_test_tt_main all_tests
+let () = run_test_tt_main all_tests
```
Comment (Dudeldu) [inline unittest/mainTest.ml]: add endline at end of file (undo this change)

File: src/cdomains/floatDomain.ml [hunk 6]
```diff
@@ -37,130 +38,210 @@ module type FloatDomainBase = sig
...
+      match op1, op2 with
+      | Some (l1, h1), Some (l2, h2) -> 
+        if l1 >= h2 then (1, 1)
+        else if h1 < l2 then (0, 0) 
+        else (0, 1)
+      | _ -> (0, 1)
+    in
+    IntDomain.IntDomTuple.of_interval IBool (Big_int_Z.big_int_of_int a, Big_int_Z.big_int_of_int b)
+
+  let eq op1 op2 =
```
Comment (Dudeldu) [inline src/cdomains/floatDomain.ml]: `eq` could probably also be implemented as `not neq` or vice versa. I see that this version is more expressive, however i would prefer less code to maintain and use a few negations where possible. But this is open to you, maybe also better to postpone this until we have all edge-cases covered with test, in order to quickly find out, whether "simplifications" don't break anything.

**Review by Dudeldu — COMMENTED** [review]
I like the idea to test our float domain with random values, however there are a few edge cases, which should always be testet (e.g. DBL_MAX + 1 -> Infinity, this would have pointed us to the missing norm after the add). In general i guess a good follow up task would be to add some tests for those cases specifically.

**Comment by Dudeldu** [comment]
Also, i was not able to install the `round` package with a similar error message as the pipeline

File: src/cdomains/floatDomain.ml (hunk 2 above)
Comment (FelixKrayer) [reply src/cdomains/floatDomain.ml]: I first removed the helper method, because I thought it was a bit confusing. However, I see the benefit and re-added it, but moved it into the FloatInterval module to make clearer where it belongs

**Comment by FelixKrayer** [comment]
About the build fail because of round: For me locally it works and I am unfortunately not able to reproduce the error. However, since we change this to our own implementation I am not really motivated to fix it. I hope we can replace it with our own implementation soon.
If you need to add the round module locally you can do so by running `opam pin add -y round https://github.com/tyabu12/ocaml-round.git`

**Comment by FelixKrayer** [comment]
Also somehow the commit history is really messed up. I tried to fix it (at some point there are two parallel FloatDomainImpl branches with the same commit) but was not successful yet as I am also not very knowledgeable at that point. Feel free to help me there if you know what to do

File: unittest/mainTest.ml (hunk 5 above)
Comment (FelixKrayer) [reply unittest/mainTest.ml]: Well I reverted it, but somehow it got worse

File: src/cdomains/floatDomain.ml [hunk 7]
```diff
@@ -1,284 +1,356 @@
...
+
+  let leq v1 v2 = 
+    match v1, v2 with
+    | _, None -> true
+    | None, Some _ -> false
+    | Some (l1, h1), Some (l2, h2) -> l1 >= l2 && h1 <= h2
+
+  let join v1 v2 = 
+    match v1, v2 with
+    | None, v | v, None -> None
```
Comment (Dudeldu) [inline src/cdomains/floatDomain.ml]: Wildcard matcher instead of `v`

File: src/cdomains/floatDomain.ml [hunk 8]
```diff
@@ -1,284 +1,356 @@
...
+      let div2d = div Down l1 h2 in
+      let div3d = div Down h1 l2 in
+      let div4d = div Down h1 h2 in
+      let high = max (max (max div1u div2u) div3u) div4u in
+      let low = min (min (min div1d div2d) div3d) div4d in
+      Some (low, high)
+
+
+  let eval_lt (l1, h1) (l2, h2) = 
+    if h1 < l2 then (1, 1)
```
Comment (Dudeldu) [inline src/cdomains/floatDomain.ml]: Maybe introduce three constants for `True = (1, 1)`, `False =(0,0)` and `Maybe = (0, 1)`

**Review by Dudeldu — DISMISSED** [review]
Looks good, no further required changes from my side, only ideas ;)

**Comment by Dudeldu** [comment]
For your git issues, maybe some useful commands, with which you should be able to solve all your issues (otherwise we can fix them together):
- `git rebase -i <commit or HEAD~x>` this allows you picking every single commit. One (drastic) cleanup is, to make all of your commits a fixup of the previous one, which results in one commit in the end. But you can also play around with reordering or leaving out commits.
- `git diff origin/master -- <file>` shows your changes in a particular file compared to the current master
- `git checkout origin/master -- <file>` resets a file to the state of the current master
- `git reset --hard origin/FloatDomainImpl` whenever you messed something up, this brings you back to the remote state of your branch (maybe requires `git fetch origin FloatDomainImpl` to fetch the current state again before)

File: goblint.opam.locked [hunk 9]
```diff
@@ -126,4 +127,8 @@ pin-depends: [
     "ppx_deriving.5.2.1"
     "git+https://github.com/ocaml-ppx/ppx_deriving.git#0a89b619f94cbbfc3b0fb3255ab4fe5bc77d32d6"
   ]
+  [
+    "round.0.1"
+    "git+https://github.com/brgr/ocaml-round.git#2dfe43050bfdac885607b6697c19c2ed913d5be4"
```
Comment (brgr) [inline goblint.opam.locked:132]: Maybe add a TODO comment here above that we (1) should replace this with our own implementation and (2) in the unlikely case that we keep it, that we move this repo then to the goblint organization, I think that would be better.

Also, did you try `make setup` at your side and saw that it worked? Also, @Dudeldu could you maybe also try that to see if this now works for you as well? For me, it works now: The culprit was the .opam file of the `round` repository, as that file had an outdated way of building it, which then broke when it was performed on my PC. It should be changed now by how it is done nowadays (e.g. goblint-cil does it roughly the same way)

Comment (Dudeldu) [reply goblint.opam.locked:132]: The current setup works for me out-of-the-box as well. @iheathrow When is your version ready?

**Review by brgr — DISMISSED** [review]
Looks good to me :)

**Review by Dudeldu — DISMISSED** [review]

**Comment by FelixKrayer** [comment]
I have now cleaned up the history so now we have 3 clean commits. Please approve again, nothing has changed in the overall, so it should be fine
(indentation only fails because history changed and it is confused because the commit it thinks is the previous one does not exist anymore)

File: goblint.opam.locked (hunk 9 above)
Comment (FelixKrayer) [reply goblint.opam.locked:132]: Added a TODO at least in the Trello board (replace ocaml-round) so we don't forget

**Review by Dudeldu — APPROVED** [review]

**Review by brgr — APPROVED** [review]
